include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+ * headers are available, the host must create a new ContextBase object to manage the new stream and
+ * call onRequestHeaders() on that object which will cause a corresponding Context to be allocated
+ * in the VM which will receive the proxy_on_context_create and proxy_on_request_headers calls.
+ *
+ * 3) For testing and instrumentation the methods of ContextBase can be replaces or augmented.
+ */
+class ContextBase {
+public:
+  ContextBase();                                                   // Testing.
+  ContextBase(WasmBase *wasm);                                     // Vm Context.
+  ContextBase(WasmBase *wasm, std::shared_ptr<PluginBase> plugin); // Root Context.
+  ContextBase(WasmBase *wasm, uint32_t root_context_id,
+              std::shared_ptr<PluginBase> plugin); // Stream context.
+  virtual ~ContextBase();
+
+  WasmBase *wasm() const { return wasm_; }
+  uint32_t id() const { return id_; }
+  bool isVmContext() { return id_ == 0; }
+  bool isRootContext() { return root_context_id_ == 0; }
+  ContextBase *root_context() { return root_context_; }
+  string_view root_id() const { return plugin_->root_id_; }
+  string_view log_prefix() const { return plugin_->log_prefix(); }
+  WasmVm *wasmVm() const;
+
+  //
+  // Calls into the VM.
+  //
+
+  // Root Context
+
+  /**
+   * Call on a Root Context when a VM first starts up.
+   * @param plugin is the plugin which caused the VM to be started.
+   */
+  virtual bool onStart(std::shared_ptr<PluginBase> plugin);
+
+  /**
+   * Call on a Root Context when a plugin is configured or reconfigured dyanmically.
+   * @param plugin is the plugin which is being configured or reconfigured.
+   */
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin);
+
+  // Context
+
+  /**
+   * Call on a host Context to create a corresponding Context in the VM.  Note:
+   * onNetworkNewConnection and onRequestHeaders() call onCreate().
+   * @param parent_context_id is the parent Context id for the context being created.  For a Network
+   * stream Context, this will be a Root Context id.
+   */
+  virtual void onCreate(uint32_t parent_context_id);
+
+  // Network
+
+  /**
+   * Call on a stream Context to indicate that a new network connection has been been created.
+   * Calls onStart().
+   */
+  virtual ProxyAction onNetworkNewConnection() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+  /**
+   * Call on a stream Context to indicate that data has arrived from downstream (e.g. on the
+   * incoming port that was accepted and for which the proxy is acting as a server).
+   * @param data_length the amount of data in the buffer.
+   * @param end_of_stream is true if no more data will be arriving.
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onDownstreamData(int /* data_length */, bool /* end_of_stream */) {
```

Comment:
Would prefer unsigned types for things that can't be negative like length. Applies elsewhere.

Reply:
Done.